### DISCUSSION TITLE
[RFC] `next lint` should accept file paths directly, without `--dir`/`--file` flags

### DISCUSSION THREAD

**Discussion by stefanprobst** [vercel/next.js#30871] [Ideas]
### Describe the feature you'd like to request

ESLint allows directly passing file paths, which should be linted. However, `next lint` requires to pass file paths either via `--file` or `--dir` flag.

The reason for this is that `next lint` accepts a `baseDir` as first argument. This is currently [not documented in the `next lint` docs](https://nextjs.org/docs/basic-features/eslint), but only in [the CLI help output](https://github.com/vercel/next.js/blob/387d6564aecc3c009fc652bb207043469687ec97/packages/next/cli/next-lint.ts#L89-L140). `next lint` also [accepts a `--base-dir` flag](https://github.com/vercel/next.js/blob/387d6564aecc3c009fc652bb207043469687ec97/packages/next/cli/next-lint.ts#L42), which currently does not seem to do anything.

`next lint` not accepting file paths directly leads to some awkward patterns, for example when used with `lint-staged`: see [this example in the docs](https://nextjs.org/docs/basic-features/eslint#lint-staged), which requires a separate `.js` config for `lint-staged`, and will not work when configured via `package.json` field.

### Describe the solution you'd like

`next lint` should directly accept file paths, like ESLint does.

Since this would currently be a breaking change (see also the previous attempt at #26777, which did not make it into Next.js 12), I propose to add a new `--paths` option, will would switch off reading a baseDir from the first argument, and instead accept file paths directly:

```
# currently
next lint --dir pages/ --dir components/ --dir lib/
next lint . --dir pages/ --dir components/ --dir lib/

# suggestion
next lint --paths pages/ components/ lib/
next lint --base-dir . --paths pages/ components/ lib/
```

Adding `--paths` would be a non-breaking change. For the next major version, it could be considered to drop reading the baseDir from the first argument, and always require passing it via `--base-dir`.

These are the suggested changes: https://github.com/vercel/next.js/compare/canary...stefanprobst:feat/eslint-file-paths

### Describe alternatives you've considered

n/a

**Comment by stefanprobst** [comment]
@timneutkens could you convert this to a discussion please? thanks!

**Comment by housseindjirdeh** [comment]
Thanks for logging this! Overall, I agree that the current behavior of passing in files and directories through different flags is a little bit wonky, so I'm all in favor of improving it :)

> I propose to add a new `--paths` option

I like it! That sounds great to me, and looking at your suggested changes I'm glad you've kept support for `--file` and `--dir` (for the meantime at least) to not introduce any breaking changes immediately. 

> For the next major version, it could be considered to drop reading the baseDir from the first argument, and always require passing it via `--base-dir`

Hmm if I recall correctly, I was planning to introduce `--base-dir` as part of the changes within #26777 and it was all reverted due to it being a breaking change. Why do we need to re-introduce `--base-dir` if we are including a new `--paths` flag that can work with the current behavior?

**Reply by stefanprobst** [reply]
> Why do we need to re-introduce --base-dir if we are including a new --paths flag that can work with the current behavior?

if i understand the current behavior correctly, a custom base dir can be provided without `--` flag, and file/folder require the `--file`/`--dir` flags. the `--paths` proposal would sort of switch this behavior around: everything without flags is interpreted as file/folder, and if a user wants to provide a custom base dir, this can be done via `--basedir` option.

**Reply by housseindjirdeh** [reply]
Ah I understand, this proposal introduces both `--path` and `--basedir` :)

That makes sense to me, and although I want to say why not just remove the `--paths` flag entirely, this brings us back to the same problem I had in #26777 where a breaking change gets introduced. This proposal seems like a solid middle ground!

> For the next major version, it could be considered to drop reading the baseDir from the first argument, and always require passing it via `--base-dir`

My only concern with this is that if we end up doing this, then I think it ultimately makes sense to not consider the `--paths` flag altogether and just aim for this directly in the next major version? I suggest that either we introduce `--paths` with the end goal of that being the stable API that sticks for the long run, or aim to introduce a breaking change to remove all file/dir flags altogether. I'm okay with either approach, but also know that the latter may take some convincing and would require some work in order to make the change as little "breaking" as possible (update the migration guide, show warnings to developers to switch their syntax beforehand, etc...)

What do you think?

**Reply by stefanprobst** [reply]
hmm, long-term, i think it's best to align `next lint` with the behavior of the ESLint CLI, i.e. accept files and folders without `--` flags. so i wouldn't want this propsal as a long term solution, it was more intended as a stop-gap until a next major version (since we missed the v12 window for a breaking change) which keeps current behavior but optionally allows providing files/folders directly. however i think you're probably right that it might be confusing to introduce optional alternative syntax only to remove it again for v13. :+1: any idea what the timeframe would be for a next major version?

**Reply by housseindjirdeh** [reply]
> i think it's best to align next lint with the behavior of the ESLint CLI, i.e. accept files and folders without -- flags

Yep, and I agree!

> so i wouldn't want this propsal as a long term solution, it was more intended as a stop-gap until a next major version (since we missed the v12 window for a breaking change) which keeps current behavior but optionally allows providing files/folders directly. however i think you're probably right that it might be confusing to introduce optional alternative syntax only to remove it again for v13

Yeah totally see your point in at least trying to simplify the syntax before we end up introducing the breaking change, but I just worry if that's too much churn and developers may wonder why the API is changing so much.

> any idea what the timeframe would be for a next major version?

I'm not entirely sure, but I would assume soon after the new year.

CC @ijjk for a second opinion on all of this :)
